Remove stack from source

vcua-mobify opened the pull request:
GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000KAC67YAH/view

# Description

Stack trace data and SSR options are visible on error pages when viewing page source. This change stops this data from being displayed in page source.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [X] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- npm ci at root to rebuild sdk
- Overwrite getProps method in app/components/_app.jsx inside to throw an error.
```
App.getProps = () => {
  throw new Error('this is an error')
}

```
# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vcua-mobify committed:
Starting release process for 1.2.0

kevinxh committed:
Merge homepage fix into release-1.2.x branch (#212)

* Remove Einstein from Homepage (#208)

* Hide conditionally the Einstein section

* Fix Warning on Error page

* Remove Einstein section

* Build translations

* Add git2gus config (#210)

* add git2gus config

Co-authored-by: Adam Raya <adamraya@users.noreply.github.com>
Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vcua-mobify committed:
Version 1.2.0-alpha.1

vcua-mobify committed:
Version 1.2.0-alpha.1 (#213)

* Version 1.2.0-alpha.1

* Update changelog for 1.2.0

vcua-mobify committed:
Merge branch 'release-1.2.x' of github.com:SalesforceCommerceCloud/pwa-kit into release-1.2.x

vcua-mobify committed:
Merge branch 'release-1.2.x' into develop

vcua-mobify committed:
Begin development on 1.3.0

vcua-mobify committed:
Merge branch 'develop' of github.com:SalesforceCommerceCloud/pwa-kit into develop

vcua-mobify committed:
Remove error and state from windowGlobals

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -266,11 +266,7 @@ const renderApp = (args) => {
-        __ERROR__: error,
-        // `window.Progressive` has a long history at Mobify and some
-        // client-side code depends on it. Maintain its name out of tradition.
-        Progressive: getWindowProgressive(req, res)
```
what's the reason behind removing `window.Progressive`? I don't believe this is publicly documented, but this is a breaking change...

vcua-mobify replied:
PR has been updated to keep window.Progressive for now.

vcua-mobify committed:
Only hide error stack trace in window globals for production

vcua-mobify commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -267,7 +269,7 @@ const renderApp = (args) => {
         __DEVICE_TYPE__: deviceType,
         __PRELOADED_STATE__: appState,
-        __ERROR__: error,
+        __ERROR__: isProduction ? undefined : error,
```
This currently sets error to undefined if the deploy target is 'production'. Using undefined here produces a 'cannot read properties of undefined' error somewhere so would using an empty string be better?

olibrook replied:
Ben asked me for an opinion! Some pointers:

1. The `process.env.DEPLOY_TARGET` var is unfortunately just a name. You can't use it reliably to find out what a production environment is on MrT, but a solution is in the works ([massive internal thread here], if you're curious).
2. Given (1), and that this is a security issue, I would probably base this behaviour on [`isRemote`]. That would mean you can't see stack traces on any deployed instance.
3. If, in future, you really want to show these stack traces on non-production deployed instances, reach out to the MrT team.
4. Don't ditch the error object entirely – we just want to get rid of the stack trace. But I think it comes with a status and a message that you would still want to show!

[massive internal thread here]: https://mobify.slack.com/archives/C8YDDMKFZ/p1637878095039600
[`isRemote`]: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/59b6a392f51ec7f7336e12434238169f1b27eebb/packages/pwa-kit-react-sdk/src/utils/ssr-server.js#L37

olibrook replied:
@vcua-mobify Seems like there's a bit of confusion about whether you even wanted to use "production" targets here at all! https://mobify.slack.com/archives/C8YDDMKFZ/p1637956473109800

bendvc replied:
Good catch @olibrook about not throwing the baby out with the bath water. We could probably do something like:

```
if (error && isRemote()) {
    delete error.stack
}
```

bendvc replied:
@vcua-mobify ☝️  has not been resolved (point number 4)

vcua-mobify replied:
Point #4 has been addressed. Only the stack is removed in remote environments.

vcua-mobify commented:
Some test environments:

Prod: https://support-test-project-production.mobify-storefront.com/
Not Prod: https://support-test-project-vincent-test-env.mobify-storefront.com/

Run ` console.log(__ERROR__.stack)` in the browser console to test

noahadams commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -256,6 +256,8 @@ const renderApp = (args) => {
 
     const helmet = Helmet.renderStatic()
 
+    const isProduction = process.env.DEPLOY_TARGET === 'production'
```
Arguably you would not want stack traces for staging, or any other environment that might contain keys or other sensitive data in the stack trace, not just production....

bendvc replied:
```
const isProduction = process.env.NODE_ENV === 'production'
```

bendvc replied:
As @olibrook mentioned below, you could also use `isRemote` it might make mocking easier and makes it semantically more correct in the sense that we are going to want this behaviour for all deployed environments (remote might sound better than using production).

olibrook replied:
@bendvc That is a much weaker protection – if you want a guarantee, you want to do `isRemote()`. Personally, I'd avoid `NODE_ENV` because the teams are already pretty confused on exactly what we want to target here.

`NODE_ENV` means _"a developer built this bundle with `NODE_ENV` set to production on their machine"_. They can then choose to deploy that bundle to any target they want. And that would allow people to reveal those stack-traces on a legitimate production site, which you're trying to prevent.

Using `isRemote()` is correct, in all cases - you'd see stack-traces on a developer machine and never see them on any bundle deployed to the Runtime.

noahadams replied:
@vcua-mobify I have updated the GUS ticket for this to better capture that stack trace data should not be shown when the application is running somewhere publicly accessible, as opposed to in managed runtime environments named "production".

Semantically, we probably want a flag labelled something like "includeStackTraceData" that defaults to false, and is set true when we know you are running in development.

vcua-mobify replied:
PR has been updated to use `isRemote()` and includes a test.

vcua-mobify committed:
Do not show stack for any remote environment

bendvc committed:
PR review changes

1. Don't squash the error object entirely
2. Iterate tests 2 times, one remote = true another remote = false

vcua-mobify commented:
@bendvc and I met to go over this PR and made some additional changes to remove only the stack but preserve the rest of the error object in remote environments.

vcua-mobify committed:
Update CHANGELOG.md

vcua-mobify committed:
Merge branch 'develop' into remove-stack-from-source

noahadams commented:
@vcua-mobify It still feels like the condition is a bit too open -- I could have been more clear, but it feels like to provide a secure default we would want to only show a trace when we know we're running in the dev server, or including stack traces has been explicitly configured on. I think it's important to consider this problem in the absence of Managed Runtime -- How can we provide a secure default to someone who might try to run this on their own server or on Heroku?

To be clear - printing stack traces in responses is a security problem and is only cool in places where we know it is not going to leak information. Maybe give this a read: https://owasp.org/Top10/A05_2021-Security_Misconfiguration/

vcua-mobify commented:
@noahadams I think the condition being too open is rooted within our current implementation of `isRemote()` since it only considers Managed Runtime as a remote environment. To properly address your concern, we will need to re-implement isRemote() so it also appropriately includes non MRT remote environments and that is out of scope for this change.

vcua-mobify committed:
Merge branch 'develop' into remove-stack-from-source
